src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java

```diff
@@ -970,6 +979,407 @@ public void recursivelyDelete(final Project project, final boolean commitIndex)
         }
     }
 
+    /**
+     * Deletes a list of Projects (identified by their UUIDs) and all objects dependent on them.
+     * @param uuids the UUIDs of the Projects to delete
+     *
+     * NB: if ON DELETE rules had been set up, this would be as simple as "delete from project where ..."
```

Comment:
I asked myself the same. Virtually no record that is associated with a project is of any use when the project is gone (exception perhaps being the project parent-child relationship).

We'll change that in v5 for sure. Changing this in v4 will be a bit challenging since we'd need to do re-create FKs manually for all supported databases. But it's not impossible. 

The DataNucleus ORM has a way to define cascading deletes on FKs, so it would automatically apply for new deployments: https://www.datanucleus.org/products/accessplatform_6_0/jdo/mapping.html#fk 

If you'd like to look into this, I think it would make your current implementation even more efficient.